### PR TITLE
[FEATURE] Alignment Configuration: Replace using_score_type tag for align_cfg::output with align_cfg::score_type

### PR DESCRIPTION
**Pull request by Irallia**
Resolves https://github.com/seqan/product_backlog/issues/179

This PR is a follow up of https://github.com/seqan/seqan3/pull/2024 and removes the seqan3::align_cfg::result completely.

### CODE REVIEW

File: include/seqan3/alignment/configuration/align_config_score_type.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+/*!\brief Tag to configure the score type of the alignment algorithm.
+ * \ingroup alignment_configuration
+ */
+// template <arithmetic score_t>
+struct score_type_tag : public pipeable_config_element<score_type_tag>
+{
+    //!\privatesection
+    //!\brief Internal id to check for consistent configuration settings.
+    static constexpr detail::align_config_id id{detail::align_config_id::score_type};
+};
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_score_type.hpp:33]:
```suggestion
template <arithmetic score_t = int32_t>
struct score_type_tag : public pipeable_config_element<score_type_tag>
{
    //!\brief The selected score type.
    using score_type = score_t;

    //!\privatesection
    //!\brief Internal id to check for consistent configuration settings.
    static constexpr detail::align_config_id id{detail::align_config_id::score_type};
};
```

File: include/seqan3/alignment/configuration/align_config_score_type.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * This option configures the score type of the alignment algorithm.
+ * By default, the alignment algorithm will only compute the score with score type `int32_t`.
+ *
+ * ### Example
+ *
+ * \include test/snippet/alignment/configuration/align_cfg_score_type.cpp
+ */
+template <arithmetic score_t = int32_t>
+inline constexpr score_type_tag score_type{};
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_score_type.hpp:50]:
```suggestion
template <arithmetic score_t>
inline constexpr score_type_tag<score_t> score_type{};
```

File: test/snippet/alignment/configuration/align_cfg_score_type.cpp [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#include <seqan3/alignment/configuration/align_config_score_type.hpp>
+
+int main()
+{
+    // Compute only the score.
+    seqan3::configuration cfg = seqan3::align_cfg::score_type<>;
```
Comment (rrahn) [inline test/snippet/alignment/configuration/align_cfg_score_type.cpp:6]:
```suggestion
    seqan3::configuration cfg = seqan3::align_cfg::score_type<int16_t>; // Now the alignment computes 16 bit integers.
    seqan3::configuration cfg = seqan3::align_cfg::score_type<float>;  // Now the alignment computes float scores.
```

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 4]
```diff
@@ -262,10 +240,16 @@ struct alignment_configurator
     static constexpr auto configure(config_t const & cfg)
     {
 
-        if constexpr (!config_t::template exists<align_cfg::result>())
+        // If no output is set, everything will set by default.
```
Comment (simonsasse) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp:243]: Not sure, if you forgot a 'be' or if I'm not familiar with this way of saying it :)
```suggestion
        // If no output is set, everything will be set by default.
```

File: include/seqan3/alignment/configuration/detail.hpp [hunk 5]
```diff
@@ -74,31 +73,29 @@ inline constexpr std::array<std::array<bool, static_cast<uint8_t>(align_config_i
...
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1}, //  9: output_begin_position
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1}, // 10: output_end_position
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1}, // 11: output_sequence1_id
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1}, // 12: output_sequence2_id
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1}, // 13: output_score
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1}, // 14: parallel
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1}, // 16: result_type
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1}, // 17: score_type
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1}, // 18: scoring
+        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0}  // 19: vectorised
```
Comment (simonsasse) [inline include/seqan3/alignment/configuration/detail.hpp:98]:
```suggestion
        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1}, // 15: result_type
        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1}, // 16: score_type
        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1}, // 17: scoring
        { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0}  // 18: vectorised
```

**Review by simonsasse — COMMENTED** [review]
Just tiny things 🧐💻

**Review by simonsasse — APPROVED** [review]
Thanks, looks good to me :)

File: test/unit/alignment/configuration/align_config_score_type_test.cpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    EXPECT_TRUE((std::same_as<decltype(seqan3::align_cfg::score_type<int16_t>)::score_type, int16_t>));
+    EXPECT_TRUE((std::same_as<decltype(seqan3::align_cfg::score_type<float>)::score_type, float>));
+    EXPECT_TRUE((std::same_as<decltype(seqan3::align_cfg::score_type<double>)::score_type, double>));
+}
+
+TEST(align_config_score_type, score_type_exists)
+{
+    seqan3::configuration cfg = seqan3::align_cfg::score_type<double>;
+    EXPECT_TRUE(cfg.exists<seqan3::align_cfg::score_type_tag<double>>());
```
Comment (rrahn) [inline test/unit/alignment/configuration/align_config_score_type_test.cpp:40]:
```suggestion
    EXPECT_TRUE(cfg.exists<seqan3::align_cfg::score_type_tag<double>>());
    EXPECT_TRUE(cfg.exists<seqan3::align_cfg::score_type_tag>());
```

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp [hunk 7]
```diff
@@ -85,7 +85,7 @@ template <template <bool, typename...> typename edit_traits_type,
           typename align_cfg_t>
 auto edit_distance(database_t && database, query_t && query, align_cfg_t && align_cfg)
 {
-    auto align_cfg_with_score_type = align_cfg | seqan3::align_cfg::result{seqan3::with_score}; // TODO remove once score type can be configured separately.
+    auto align_cfg_with_score_type = align_cfg;
```
Comment (rrahn) [inline test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp:88]: But then this is not necessary. It does not makes sense to just copy it here. So you can work with the default align_cfg. Please update the part below to use the original align_cfg if we know already that it has an output configured.

**Review by rrahn — CHANGES_REQUESTED** [review]
Thank you very much. Some changes I'd like to have :)

File: doc/tutorial/pairwise_alignment/index.md [hunk 8]
```diff
@@ -311,8 +311,9 @@ To make the configuration easier, we added a shortcut called seqan3::align_cfg::
 The edit distance can be further refined using seqan3::align_cfg::aligned_ends to also compute a semi-global alignment
 and the seqan3::align_cfg::min_score configuration to fix an edit score (a limit of the allowed number of edits). If the
 respective alignment could not find a solution within the given error bound, the resulting score is infinity
-(corresponds to std::numeric_limits::max). Also the alignment and the front and back coordinates can be computed using
-the align_cfg::result option.
+(corresponds to std::numeric_limits::max). Also the alignment and the front and back coordinates can be computed using a
```
Comment (rrahn) [inline doc/tutorial/pairwise_alignment/index.md:314]:
```suggestion
(corresponds to std::numeric_limits::max). Also the alignment and the begin and end positions of the alignment can be computed using a
```

File: include/seqan3/alignment/configuration/align_config_score_type.hpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides configuration for alignment output.
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_score_type.hpp:9]:
```suggestion
 * \brief Provides alignment configuration seqan3::align_cfg::score_type.
```

File: include/seqan3/alignment/configuration/align_config_score_type.hpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+ * \author Lydia Buntrock <lydia.buntrock AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/configuration/detail.hpp>
+#include <seqan3/core/algorithm/configuration.hpp>
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_score_type.hpp:16]:
```suggestion
```

File: include/seqan3/alignment/configuration/align_config_score_type.hpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    static_assert(std::floating_point<score_t> || std::signed_integral<score_t>,
+                  "The selected score type must be a signed integral type or floating point type.");
+    //!\brief The selected score type.
+    using score_type = score_t;
+    //!\privatesection
+    //!\brief Internal id to check for consistent configuration settings.
+    static constexpr seqan3::detail::align_config_id id{seqan3::detail::align_config_id::score_type};
+};
+
+/*!\brief Helper variable used to configure the score type for the alignment algorithm.
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_score_type.hpp:38]:
```suggestion
/*!\brief A configuration element to set the score type used in the alignment algorithm.
```

File: include/seqan3/alignment/configuration/align_config_score_type.hpp [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+ * \ingroup alignment_configuration
+ * \relates seqan3::align_cfg::output_alignment
```
Comment (rrahn) [inline include/seqan3/alignment/configuration/align_config_score_type.hpp:40]:
```suggestion
```

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 13]
```diff
@@ -261,106 +240,97 @@ struct alignment_configurator
...
-            // Configure the alignment algorithm.
-            return std::pair{configure_scoring_scheme<function_wrapper_t>(config_with_result_type),
-                             config_with_result_type};
         }
+
+        // ----------------------------------------------------------------------------
+        // Check if invalid configuration was used.
+        // ----------------------------------------------------------------------------
+
+        // Do not allow max error configuration for alignments not computing the edit distance.
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp:327]:
```suggestion
        // Do not allow min score configuration for alignments not computing the edit distance.
```

File: include/seqan3/alignment/pairwise/alignment_result.hpp [hunk 14]
```diff
@@ -348,7 +348,7 @@ class alignment_result
      * seqan3::align_cfg::detail::debug.
      *
      * \note This function is only available if the debug mode and the alignment was requested via the alignment
-     * configuration (see seqan3::align_cfg::detail::debug and seqan3::align_cfg::result).
+     * configuration (see seqan3::align_cfg::detail::debug and seqan3::align_cfg::output_score).
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/alignment_result.hpp:351]:
```suggestion
     * configuration (see seqan3::align_cfg::detail::debug and seqan3::align_cfg::output_alignment).
```

File: test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp [hunk 15]
```diff
@@ -85,15 +85,14 @@ template <template <bool, typename...> typename edit_traits_type,
...
+    using align_cfg_with_score_type_t = decltype(align_cfg);
 
     using alignment_result_value_t =
         typename seqan3::detail::align_result_selector<database_t,
                                                        query_t,
                                                        std::remove_reference_t<align_cfg_with_score_type_t>>::type;
     using alignment_result_t = seqan3::alignment_result<alignment_result_value_t>;
-    auto align_cfg_with_result_type = align_cfg_with_score_type |
+    auto align_cfg_with_result_type = align_cfg |
                                       seqan3::align_cfg::detail::result_type<alignment_result_t>;
```
Comment (rrahn) [inline test/unit/alignment/pairwise/edit_distance_unbanded_test_template.hpp:96]:
```suggestion
    auto align_cfg_with_result_type = align_cfg | seqan3::align_cfg::detail::result_type<alignment_result_t>;
```

**Review by rrahn — CHANGES_REQUESTED** [review]
Ok, second round is done. Looks quite good now! Thanks a lot.

**Review by rrahn — APPROVED** [review]
ok looks good now. Thank you very much. Can be rebased now and shipped :shipit:

**Comment by marehr** [comment]
@Irallia Can you next time make sure that every single commit in your PRs can be compiled?

Commits in this PR
1. 9078045ca1047c9e8bf99e982171390e13032a67 does not compile
2. a8d81ce1a08feba4652175136bf84ef90e820787 does not compile
3. 9619f834b4c8286a01bd03faab546cc6d0601378 does not compile
4. a79d5d466c7b1e1a3e1d9d9bc9a11f84272a3369 does not compile
5. d0283da05e08904484ce6cc908d67d7a05fc1b05 does not compile
6. 8d0e389a0f7495706fc05b51de43650f30ec2229 does compile
7. f4e146c92f43e6d3c53452c3ba8e1c9ef40fe505 does compile

Otherwise, it makes it really hard to figure out which commit introduced a bug/regression.

**Pull request merged**